Add driver for generic HID joysticks/gamepads

valpackett opened the pull request:
Tested on an off-brand cheap USB gamepad with <16 buttons :D but I'm relatively confident in the lots-of-buttons (`BTN_TRIGGER_HAPPY`) support code.

The list of individual buttons in `struct hmap_item[]` sucks, would be nice to have a `.usage_mask` instead of `.usage` that would allow me to match all `HUP_BUTTON`s…

valpackett committed:
[hmap] pass the hmap_item to callbacks

valpackett committed:
[hgame] Add driver for generic HID joysticks/gamepads

wulf7 commented:
Thanks for hgame driver.

Could you rework it without using callbacks? Although currently callbacks look better, there are plans to implement HID usage ranges and scatter-gather maps for such a drivers

valpackett committed:
[hgame] Do not use a callback

valpackett committed:
[hgame] Add support for D-pad usages found on XBox 360 gamepads

Also add comment about testing on an XBox 360/XInput compatible gamepad.

valpackett committed:
[hgame] Add XBox 360 gamepad mapping

Using the direct number mapping with our custom descriptor
results in buttons being wrong.

For now, match using Microsoft vendor.

valpackett commented:
- Removed callback usage for regular buttons
- Added support for d-pad usages (that required a new callback though :D)
  - basically, "hat switch" is when there's one HID usage that returns a value like: 1 is up, 2 is up-right, 3 is right, …
    - Linux exposes that value directly as the X axis of a hat switch
  - "d-pad" is like 4 separate buttons
    - but Linux exposes that as two axes of a hat switch that go from -1 to 1, hence somewhat ugly callback
- Added correct button mapping for Xbox 360 controllers
  - currently matching by vendor (Microsoft), would be nice to somehow use the `USB_IFACE_SUBCLASS(UISUBCLASS_XBOX360_CONTROLLER)` knowledge from `usbhid` instead
    - interestingly, the [xpad](https://github.com/paroj/xpad/blob/master/xpad.c) driver for Linux keeps a list of vendor/device pairs that includes 3rd party vendors. (My 3rd party gamepad just pretends to be Microsoft when it's in XInput mode lol)

wulf7 commented:
Merged, thanx

> that required a new callback though
> hence somewhat ugly callback

Callbacks were added for things like this one, They should be used if translation can not be written in declarative way.

> currently matching by vendor (Microsoft), would be nice to somehow use the USB_IFACE_SUBCLASS(UISUBCLASS_XBOX360_CONTROLLER) knowledge from usbhid instead

May be create hgame.h with required constants?

valpackett commented:
> May be create hgame.h with required constants?

I mean how would I pass the info between drivers.. I guess via `hid_device_info`, which is passed as ivars

wulf7 commented:
Yes, hid_device_info is intended to pass information between layers. Do you want place quirks here too?

valpackett commented:
Yeah, I guess. (I specifically need to pass a USB-level match to a HID driver)

wulf7 commented:
Other way is a global function that compares `hid_device_info` structure with the predefined quirk list. That is what USB's usb_test_quirk (and usb_quirk(4)) do. But I never dug in to it so I do not know if it can replace passing quirks between drivers.
